File: hello-grpc-dart/lib/common/logging_config.dart
/// Logging configuration for standardized logging setup.
///
/// Provides utilities to initialize logging with standard format:
/// `[TIMESTAMP] [LEVEL] [COMPONENT] MESSAGE [CONTEXT]`
///
/// Uses Dart's logging package with dual output (console and file).
library;

import 'dart:io';

import 'package:intl/intl.dart';
import 'package:logging/logging.dart';

const String defaultLogDir = 'logs';
final DateFormat timestampFormat = DateFormat('yyyy-MM-dd HH:mm:ss.SSS');
final DateFormat fileTimestampFormat = DateFormat('yyyyMMdd_HHmmss');

/// Logging configuration helper
class LoggingConfig {
  /// Initialize logging for a component with dual output (console and file)
  ///
  /// [component] - The component name (e.g., "client", "server")
  /// [logDir] - The directory for log files (default: "logs")
  /// [enableFile] - Whether to enable file logging (default: true)
  ///
  /// Returns the configured logger instance
  static Logger initializeLogging(
    String component, {
    String logDir = defaultLogDir,
    bool enableFile = true,
  }) {
    // Create log directory
    if (enableFile) {
      final dir = Directory(logDir);
      if (!dir.existsSync()) {
        dir.createSync(recursive: true);
      }
    }

    // Get log level
    final level = getLogLevel();

    // Configure logging
    Logger.root.level = level;

    // Create log file if enabled
    File? logFile;
    if (enableFile) {
      final timestamp = fileTimestampFormat.format(DateTime.now());
      final logFileName = '${component}_$timestamp.log';
      final logFilePath = '$logDir/$logFileName';
      logFile = File(logFilePath);
    }

    // Set up log handler
    Logger.root.onRecord.listen((record) {
      final timestamp = timestampFormat.format(record.time);
      final level = record.level.name;
      final message = record.message;

      // Build context from error and stack trace
      var context = '';
      if (record.error != null || record.stackTrace != null) {
        final contextParts = <String>[];
        if (record.error != null) {
          contextParts.add('error=${record.error}');
        }
        if (contextParts.isNotEmpty) {
          context = ' [${contextParts.join(', ')}]';
        }
      }

      // Build log line
      final logLine =
          '[$timestamp] [$level] [${record.loggerName}] $message$context';

      // Output to console
      // ignore: avoid_print
      print(logLine);

      // Output to file if enabled
      if (logFile != null) {
        logFile.writeAsStringSync('$logLine\n', mode: FileMode.append);
      }

      // Add stack trace if present
      if (record.stackTrace != null) {
        final stackTrace = record.stackTrace.toString();
        // ignore: avoid_print
        print(stackTrace);
        if (logFile != null) {
          logFile.writeAsStringSync('$stackTrace\n', mode: FileMode.append);
        }
      }
    });

    // Create logger
    final logger = Logger(component)
      ..info('Logging initialized for component: $component');
    if (enableFile) {
      logger.info('Log directory: $logDir');
    }

    return logger;
  }

  /// Get log level from environment variable or default to INFO
  static Level getLogLevel() {
    final levelStr = Platform.environment['LOG_LEVEL']?.toUpperCase() ?? 'INFO';

    switch (levelStr) {
      case 'ALL':
        return Level.ALL;
      case 'FINEST':
      case 'TRACE':
        return Level.FINEST;
      case 'FINER':
      case 'DEBUG':
        return Level.FINER;
      case 'FINE':
        return Level.FINE;
      case 'CONFIG':
        return Level.CONFIG;
      case 'INFO':
        return Level.INFO;
      case 'WARNING':
      case 'WARN':
        return Level.WARNING;
      case 'SEVERE':
      case 'ERROR':
        return Level.SEVERE;
      case 'SHOUT':
      case 'FATAL':
        return Level.SHOUT;
      case 'OFF':
        return Level.OFF;
      default:
        return Level.INFO;
    }
  }
}

/// Convenience function to initialize logging
Logger initializeLogging(
  String component, {
  String logDir = defaultLogDir,
  bool enableFile = true,
}) {
  return LoggingConfig.initializeLogging(
    component,
    logDir: logDir,
    enableFile: enableFile,
  );
}

File: hello-grpc-ts/common/loggingConfig.ts
/**
 * Logging configuration for standardized logging setup.
 * 
 * Provides utilities to initialize logging with standard format:
 * [TIMESTAMP] [LEVEL] [COMPONENT] MESSAGE [CONTEXT]
 * 
 * Uses winston with dual output (console and file).
 */

import winston from 'winston';
import * as fs from 'fs';
import * as path from 'path';

const DEFAULT_LOG_DIR = 'logs';
const TIMESTAMP_FORMAT = 'YYYY-MM-DD HH:mm:ss.SSS';

/**
 * Custom format for standard log output
 */
const standardFormat = (component: string) => winston.format.printf(({ timestamp, level, message, ...metadata }) => {
    let log = `[${timestamp}] [${level.toUpperCase()}] [${component}] ${message}`;
    
    // Add context if present
    const levelSymbol = Symbol.for('level');
    const contextKeys = Object.keys(metadata).filter(key => 
        key !== 'timestamp' && key !== 'level' && key !== 'message' && key !== levelSymbol.toString()
    );
    
    if (contextKeys.length > 0) {
        const context = contextKeys.map(key => `${key}=${metadata[key]}`).join(', ');
        log += ` [${context}]`;
    }
    
    return log;
});

/**
 * Get log level from environment variable or default to 'info'
 */
function getLogLevel(): string {
    const level = process.env.LOG_LEVEL || 'info';
    return level.toLowerCase();
}

/**
 * Initialize logging for a component with dual output (console and file)
 * 
 * @param component - The component name (e.g., "client", "server")
 * @param logDir - The directory for log files (default: "logs")
 * @param enableFile - Whether to enable file logging (default: true)
 * @returns Logger instance for the component
 */
export function initializeLogging(
    component: string, 
    logDir: string = DEFAULT_LOG_DIR, 
    enableFile: boolean = true
): winston.Logger {
    // Create log directory
    if (enableFile && !fs.existsSync(logDir)) {
        fs.mkdirSync(logDir, { recursive: true });
    }

    const level = getLogLevel();
    const transports: winston.transport[] = [];

    // Console transport
    transports.push(
        new winston.transports.Console({
            level: level,
            format: winston.format.combine(
                winston.format.colorize(),
                winston.format.timestamp({ format: TIMESTAMP_FORMAT }),
                standardFormat(component)
            )
        })
    );

    // File transport
    if (enableFile) {
        const timestamp = new Date().toISOString().replace(/[:.]/g, '-').slice(0, -5);
        const logFileName = `${component}_${timestamp}.log`;
        const logFilePath = path.join(logDir, logFileName);

        transports.push(
            new winston.transports.File({
                filename: logFilePath,
                level: level,
                format: winston.format.combine(
                    winston.format.timestamp({ format: TIMESTAMP_FORMAT }),
                    standardFormat(component)
                ),
                maxsize: 100 * 1024 * 1024, // 100MB
                maxFiles: 7
            })
        );
    }

    // Create logger
    const logger = winston.createLogger({
        level: level,
        transports: transports,
        exitOnError: false
    });

    logger.info(`Logging initialized for component: ${component}`);
    if (enableFile) {
        logger.info(`Log directory: ${logDir}`);
    }

    return logger;
}

/**
 * Create a child logger with additional context
 * 
 * @param logger - Parent logger
 * @param context - Additional context to include in all log messages
 * @returns Child logger with context
 */
export function createChildLogger(logger: winston.Logger, context: Record<string, any>): winston.Logger {
    return logger.child(context);
}

export { getLogLevel };
